docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,389 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
->   }
-> })());
-> console.log(data[0]["prop-1"].id);
-> console.log(data.name);
-> ```
+```js
+struct Employee { 
+  id: str;
+  name: str;
+}
 
-</details>
+let response = http_get("/employees"); // returns something like { "items": [ { "id": "12234", "name": "bob" }, ... ] }
+let employees = Array<Employee>.from_json(response.items);
+
+for e in employees {
+  print("hello, ${e.name}, your employee id is ${id}");
+}
+```
+
+In the above example, the `http_get` function returns a `Json` object from the server that has a
+single field `items`, with a JSON array of JSON objects, each with an `id` and `name` fields.
+
+The expression `response.items` returns a `Json` array, and we use `Array<T>.from_json` to convert
+this array from `Json` to an `Array<Employee>`. Note that by default `from_json` will perform schema
+validation on the array and on each item (based on the declaration of the `Employee` struct).
+
+##### 1.1.4.1 Literals
+
+Literals can be defined using the `Json` type initializers:
+
+```js
+let json_string  = Json "hello";
+let json_number  = Json 123;
+let json_bool    = Json true;
+let json_array   = Json [ 1, 2, 3 ];
+let json_obj     = Json { boom: 123 };
+let json_mut_obj = MutJson {
+  hello: 123, 
+  world: [ 1, "cat", 3 ],       // <-- heterogenous array
+  "boom boom": { hello: 1233 }  // <-- non-symbolic key
+};
+```
+
+Every value within a `Json` array or object also has a type of `Json`.
+
+##### 1.1.4.2 JSON objects
+
+To access a field within an object, use the `.` notation:
+
+```js
+let boom: Json = json_obj.boom;
+```
+
+Trying to access a non-existent field will fail at runtime. For example:
+
+```js
+print(json_obj.boom.dude.world);
+// RUNTIME ERROR: Uncaught TypeError: Cannot read properties of undefined (reading 'world')
+```
+
+Like in JavaScript, it is also possible to access object fields using `[]`:
+
+```js
+let foo = j["my-field"].your_field["their-field"];
+```
+
+To obtain an array of all the keys within a JSON object use the `Json.keys(o)` method. 
+
+
+```js
+let j = { hello: 123, world: [ 1, 2, 3 ] };
+assert(Json.keys(j) == ["hello", "world"]);
+```
+
+To obtain an array of all the values, use `Json.values(o)`. To obtain an array of all key/value
+pairs use `Json.entries(o)` (P2):
+
+```js
+assert(Json.values(j).equals(Json [ 123, [ 1, 2, 3 ] ]));
+assert(Json.entries(j).equals(Json [
+  [ "hello", 123 ],
+  [ "world", [ 1, 2, 3 ] ]
+]));
+```
+
+> NOTE: `values()` and `entries()` return an array inside a `Json` object because at the moment we
+> cannot represent heterogenous arrays in Wing.
+
+##### 1.1.4.3 JSON arrays
+
+To access an array element, use the `[]` notation:
+
+```js
+let item2: Json = json_array[2];
+```
+
+Trying to index a value that is not an array will return JavaScript `undefined`.
+
+##### 1.1.4.4 Assignment from native types
+
+It is also possible to assign the native `str`, `num`, `bool` and `Array<T>` values and they will
+implicitly be casted to `Json`:
```

Comment:
Does this mean if a function has a `Json` parameter I can pass it a `str` and it will be upcasted (downcasted? I'm bad at directions so I have no idea which is correct 😅 ) for free?

Reply:
Yes, this should work I believe.